test/e2e/network/networking_perf.go

```diff
@@ -160,3 +172,275 @@ var _ = SIGDescribe("Networking IPerf [Experimental] [Slow] [Feature:Networking-
+				if pod.Status.Phase != v1.PodRunning {
+					return false, nil
+				}
+			}
+			return true, nil
+		})
+		framework.ExpectNoError(err, "unable to wait for client pods to come up")
+
+		// Get a reference to the server pod for later
+		serverPodList, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).List(context.TODO(), serverPodsListOptions)
+		framework.ExpectNoError(err)
+		if len(serverPodList.Items) != 1 {
+			framework.Failf("expected 1 server pod, found %d", len(serverPodList.Items))
+		}
+		serverPod := serverPodList.Items[0]
+
+		// Get a reference to the client pods for later
+		clientPodList, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).List(context.TODO(), clientPodsListOptions)
+		framework.ExpectNoError(err)
+		framework.Logf("found %d matching client pods", len(clientPodList.Items))
+		if len(clientPodList.Items) == 0 {
+			framework.Failf("found 0 matching client pods")
+		}
+
+		// Keep track of all the iperf2 results
+		nodeResults := &IPerf2NodeToNodeResults{
+			ServerNode: serverPod.Spec.NodeName,
+			Results:    map[string]*IPerf2EnhancedResults{},
+		}
+		// Step 3: iterate through the client pods one by one, running iperf2 in client mode to transfer
+		//   data to the server and back and measure bandwidth
+		for _, pod := range clientPodList.Items {
+			podName := pod.Name
+			nodeName := pod.Spec.NodeName
+			hostIP := pod.Status.HostIP
+
+			command := fmt.Sprintf(`iperf -e -p %d --reportstyle C -i 1 -c iperf2-server -T "Client on %s"`, iperf2Port, hostIP)
+			framework.Logf("attempting to run command '%s' in client pod %s (node %s)", command, podName, nodeName)
+			output := f.ExecShellInPod(podName, command)
+			framework.Logf("output from exec on client pod %s (node %s): \n%s\n", podName, nodeName, output)
+
+			results, err := ParseIPerf2EnhancedResults(output)
+			framework.ExpectNoError(err, "unable to parse iperf2 output from client pod %s (node %s)", pod.Name, nodeName)
+			nodeResults.Results[nodeName] = results
+		}
+
+		// Step 4: after collecting all the client<->server data, compile and present the results
+		/*
+			Example output:
+
+			Dec 10 09:13:05.658: INFO:                      From                       To         Bandwidth (MB/s)
+			Dec 10 09:13:05.658: INFO:                abc-worker               abc-worker                     2001
+			Dec 10 09:13:05.658: INFO:               abc-worker2               abc-worker                     1317
+			Dec 10 09:13:05.658: INFO:         abc-control-plane               abc-worker                     1267
+
+		*/
+		framework.Logf("%25s%25s%25s", "From", "To", "Bandwidth (MB/s)")
+		for nodeFrom, results := range nodeResults.Results {
+			framework.Logf("%25s%25s%25d", nodeFrom, nodeResults.ServerNode, results.Total.bandwidthMB())
+		}
+		for clientNode, results := range nodeResults.Results {
+			megabytesPerSecond := results.Total.bandwidthMB()
+			if megabytesPerSecond < iperf2BaselineBandwidthMegabytesPerSecond {
```

Comment:
there are a lot of people these days running kubernetes in raspberry and targeting IOT scenarios that I assume will have very little bandwidth, just food for thoughts

Reply:
+1 this is a tricky one, the goal is to set the bandwidth cutoff high enough that you know that the cluster's network isn't broken, but low enough that it won't report a working cluster as broken.  

Seems like a tricky balance, any suggestions how we can find that balance?

Reply:
I think that we should not target throughput (throughput is part of performance not performance) but correctness , based on that my criteria to pass the test will be :
- keep at least 90Mbps (consider a 100 mbps nic plus a 10% margin)
- less than 3 % of errors

Reply:
This is odd, I don't see error counts with the iperf `--reportstyle C` option, gonna hack around some more on this and see why that's not showing up.

Reply:
🤔 the snippet in the description show errors

Reply:
Unfortunately the CSV output doesn't include the errors, let me whip together something to parse the non-CSV output!

Reply:
indeed, surprisingly it doesn' have the same output
```
$ iperf -c localhost -e -y C
20201221235135.693,127.0.0.1,55932,127.0.0.1,5001,3,0.0-10.0,53096480768,42477116651
```

```
$ iperf -c localhost -e 
------------------------------------------------------------
Client connecting to localhost, TCP port 5001 with pid 3551429
Write buffer size:  128 KByte
TCP window size: 4.00 MByte (default)
------------------------------------------------------------
[  3] local 127.0.0.1 port 56228 connected with 127.0.0.1 port 5001 (ct=0.05 ms)
[ ID] Interval        Transfer    Bandwidth       Write/Err  Rtry     Cwnd/RTT        NetPwr
[  3] 0.00-10.00 sec  48.3 GBytes  41.5 Gbits/sec  395726/0          0       -1K/23 us  225515464.25

```

Reply:
okay put together https://github.com/kubernetes/kubernetes/pull/94015/commits/bf4874a921c258bbb701a6a76680a37d00c2435c to scrape out the error rate -- what do you think @aojea which approach is better?  The parsing code is kinda wonky and fragile but it gets the job done

Reply:
>  The parsing code is kinda wonky and fragile but it gets the job done
indeed, I think that since you implemented both, first we use the CSV one and we let it soak, and we revisit after some time ... my apologies, thought it was simpler

Reply:
np, glad we tried it out both ways!  👍